vowpalwabbit/features_data.h

```diff
@@ -0,0 +1,147 @@
+#include <array>
+#include <memory>
+#include <type_traits>
+
+/*
+ * class features_data
+ * Description:
+ *   This data structure manages access and lifetime of the features data used in various VW reductions.
+ *     New types must be created with the templatized create() function call, attempting to access
+ *     uninitialized data will return a nullptr.
+ *
+ *   All contained data types MUST instantiate a `clear()` function that readies the object for reuse.
+ *
+ *   Calling create() on an existing element will clear and overwrite the existing data
+ *
+ * Member functions:
+ *   T* create(constructor_parameters);
+ *   T* get();
+ *   void clear();
+ *
+ * Data structure usage:
+ *   Add new data structure to the available types:
+ *     Add a new line to the TYPE_MAPPINGS macro
+ *
+ *   Create an instance of a data structure:
+ *     features_data fd;
+ *     fd.create<data_type>(constructor_parameters);
+ *
+ *   Accessing an instance of existing data:
+ *     auto data = fd.get<data_type>();
+ */
+
+/////////////////////////////////////////////
+// sample placeholder structs. For demonstrative purposes only
```

Comment:
These placeholders will need to be removed before this goes out for final review. They're here to demonstrate how the TYPE_MAPPINGS macro will need to look